### PR TITLE
Implement the snapshot cache

### PR DESCRIPTION
**Pull request by steeling**
This PR makes the final tweaks to implement and migrate to the snapshot cache. A few notes on how this is accomplished:

1. The wokerpool is further simplified to just take a single function, while the logging is moved to the `scheduleUpdate` method. Doing this helped keep some of the tests that previously expected us to be able to "send" a response, but with the new cache we don't directly send responses.
3. Proxy version is no longer stored on the proxy, neither are "subscribed resources", since we always recalculate the whole set.
4. Switch the http_connection_manager name back to the wellknown name due to https://github.com/envoyproxy/go-control-plane/pull/589
5. All of the request objects are no longer passed into the NewResponse methods, so no need to deal with those edge cases. That allows us to remove a lot of code, and a the test cases for that code.
6. The node ID for the sidecar is changed from the XDS cert to the UUID, which is considered more stable due to the ability to rotate trust domains.

Additional testing:

1. Deploy demo
2. Query controller logs and verify the snapshot cache reaches a steady state, with no further updates.
3. Switch to permissive mode.
4. Verify all sidecars are updated with all other pods
5. Query controller logs and verify the snapshot cache reaches a steady state.

Logs uploaded
[snapshot-controller.log](https://github.com/openservicemesh/osm/files/9509131/snapshot-controller.log)

### CODE REVIEW

**Review by keithmattix — APPROVED** [review]
Code looks good; this review is just a peppering of questions from a snapshot cache noob

File: pkg/envoy/ads/cache_stream.go [hunk 1]
```diff
@@ -3,32 +3,140 @@ package ads
...
+		if s.catalog.GetMeshConfig().Spec.Observability.EnableDebugServer {
+			s.trackXDSLog(proxy.GetName(), typeURI)
+		}
+
+		// Invoke XDS handler
+		startedAt := time.Now()
+		resources, err := handler(s.catalog, proxy, s.certManager, s.proxyRegistry)
+		xdsPathTimeTrack(startedAt, typeURI, proxy, err == nil)
+		if err != nil {
+			log.Error().Err(err).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrGeneratingReqResource)).Str("proxy", proxy.String()).
```
[the remark this answers is not in the document]
Comment (steeling) [reply pkg/envoy/ads/cache_stream.go:40]: i think it's supposed to stand for "requested" resource, but agree it's confusing. I'll take a note to revisit this

File: pkg/envoy/ads/grpc.go [hunk 2]
```diff
@@ -187,28 +180,3 @@ func (s *GRPCServer) GrpcServe(ctx context.Context, cancel context.CancelFunc, l
 	}()
 	return nil
 }
-
-func receive(requests chan *xds_discovery.DiscoveryRequest, server *xds_discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer, proxy *envoy.Proxy, quit chan struct{}) {
```
[the remark this answers is not in the document]
Comment (steeling) [reply pkg/envoy/ads/grpc.go:191]: the snapshot cache! instead of us managing request state, we punt that all off to envoy's snapshot cache

File: pkg/envoy/ads/response_benchmark_test.go [hunk 3]
```diff
@@ -164,13 +158,11 @@ func BenchmarkSendXDSResponse(b *testing.B) {
...
 
-			// Set subscribed resources
-			proxy.SetSubscribedResources(xdsType, mapset.NewSetWith("service-cert:default/bookstore", "root-cert-for-mtls-inbound"))
-
 			b.ResetTimer()
 			b.StartTimer()
 			for i := 0; i < b.N; i++ {
-				if err := adsServer.sendResponse(proxy, &server, nil, xdsType); err != nil {
+				handler := adsServer.xdsHandlers[xdsType]
+				if _, err := handler(adsServer.catalog, proxy, adsServer.certManager, adsServer.proxyRegistry); err != nil {
```
[the remark this answers is not in the document]
Comment (steeling) [reply pkg/envoy/ads/response_benchmark_test.go:165]: we're no longer sending the configs ourselves, instead we populate the snapshot cache which sends the configs. So this is the best we can do to benchmark

File: pkg/envoy/ads/types.go [hunk 4]
```diff
@@ -27,7 +26,7 @@ var (
 type Server struct {
 	catalog       catalog.MeshCataloger
 	proxyRegistry *registry.ProxyRegistry
-	xdsHandlers   map[envoy.TypeURI]func(catalog.MeshCataloger, *envoy.Proxy, *xds_discovery.DiscoveryRequest, *certificate.Manager, *registry.ProxyRegistry) ([]types.Resource, error)
+	xdsHandlers   map[envoy.TypeURI]func(catalog.MeshCataloger, *envoy.Proxy, *certificate.Manager, *registry.ProxyRegistry) ([]types.Resource, error)
```
[the remark this answers is not in the document]
Comment (steeling) [reply pkg/envoy/ads/types.go:29]: more obfuscation by the snapshot cache. We just fill the configs, and we let the cache handle the rest

File: pkg/envoy/proxy.go [hunk 5]
```diff
@@ -113,48 +101,6 @@ func (p *Proxy) StatsHeaders() map[string]string {
 	}
 }
 
-// SetLastAppliedVersion records the version of the given Envoy proxy that was last acknowledged.
-func (p *Proxy) SetLastAppliedVersion(typeURI TypeURI, version uint64) {
```
[the remark this answers is not in the document]
Comment (steeling) [reply pkg/envoy/proxy.go:117]: yup that's correct!

File: pkg/workerpool/workerpool.go [hunk 6]
```diff
@@ -78,9 +65,8 @@ func NewWorkerPool(nWorkers int) *WorkerPool {
 
 // AddJob posts the job on a worker queue
 // Uses Hash underneath to choose worker to post the job to
-func (wp *WorkerPool) AddJob(job Job) chan struct{} {
-	wp.jobs <- job
-	return job.GetDoneCh()
+func (wp *WorkerPool) AddJob(f func()) {
+	wp.jobs <- f
```
[the remark this answers is not in the document]
Comment (steeling) [reply pkg/workerpool/workerpool.go:69]: they're fire and forget, but the one place they are used I re-added the synch logic, so we do end up waiting for them to complete.

This made it easier to switch the generation of configs from the current sending of configs (for testing purposes)

File: pkg/envoy/ads/cache_stream.go [hunk 7]
```diff
@@ -3,32 +3,140 @@ package ads
...
+				Msgf("Error generating response for typeURI: %s", typeURI.Short())
+			xdsPathTimeTrack(time.Now(), envoy.TypeADS, proxy, false)
+			return nil, err
+		}
+		// resources above ^^
```
[the remark this answers is not in the document]
Comment (steeling) [reply pkg/envoy/ads/cache_stream.go]: yup, fixed

File: pkg/injector/bootstrap.go [hunk 8]
```diff
@@ -30,16 +31,16 @@ func (wh *mutatingWebhook) createEnvoyBootstrapFromExisting(newBootstrapSecretNa
...
 
-	return wh.marshalAndSaveBootstrap(newBootstrapSecretName, namespace, config, cert)
+	return wh.marshalAndSaveBootstrap(bootstrapConfigName(proxyUUID), namespace, config, cert)
 }
 
-func (wh *mutatingWebhook) createEnvoyBootstrapConfig(name, namespace, osmNamespace string, cert *certificate.Certificate, originalHealthProbes models.HealthProbes) (*corev1.Secret, error) {
+func (wh *mutatingWebhook) createEnvoyBootstrapConfig(proxyUUID uuid.UUID, namespace string, cert *certificate.Certificate, originalHealthProbes models.HealthProbes) (*corev1.Secret, error) {
 	builder := bootstrap.Builder{
-		NodeID: cert.GetCommonName().String(),
+		NodeID: proxyUUID.String(),
```
Comment (jsturtevant) [inline pkg/injector/bootstrap.go:41]: I don't know what the implications of this change would be but should this be a separate PR in case there it is need to change?  would make this current change a smaller change set too.

File: pkg/envoy/ads/cache_stream.go [hunk 9]
```diff
@@ -3,32 +3,137 @@ package ads
...
+	case chunks[3] == "":
+		return "", uuid.UUID{}, "", errInvalidCertificateCN
+	}
+
+	return envoy.ProxyKind(chunks[1]), proxyUUID, identity.New(chunks[2], chunks[3]), nil
+}
+
+// recordPodMetadata records pod metadata and verifies the certificate issued for this pod
+// is for the same service account as seen on the pod's service account
+func (s *Server) recordPodMetadata(p *envoy.Proxy) error {
```
Comment (jsturtevant) [inline pkg/envoy/ads/cache_stream.go:102]: guess this won't be needed since https://github.com/openservicemesh/osm/pull/5058 is open

File: pkg/envoy/ads/cache_callbacks.go [hunk 10]
```diff
@@ -70,15 +74,33 @@ func (s *Server) OnStreamOpen(ctx context.Context, streamID int64, typ string) e
...
+			t := time.Now()
+			log.Debug().Msgf("Starting update for proxy %s", proxy.String())
+
+			if err := s.update(proxy); err != nil {
+				log.Error().Err(err).Str("proxy", proxy.String()).Msg("Error generating resources for proxy")
+			}
+			log.Debug().Msgf("Update for proxy %s took took %v", proxy.String(), time.Since(t))
+			wg.Done()
+		})
+	wg.Wait()
```
Comment (jsturtevant) [inline pkg/envoy/ads/cache_callbacks.go:91]: could this cause some delays if there are other things in the queue and we are waiting for it complete?

Comment (steeling) [reply pkg/envoy/ads/cache_callbacks.go:91]: ya but it's somewhat meaningless to schedule multiple simultaneously. each schedule tries to grab a new state, so we don't know if the current job has all the state or not, or whether the new one will anyways.

We use "coalescing" in the message broker to account for this. It probably makes sense to move that coalescing here, but I'll leav that out of this PR for now.

This mimics the current setup as well.

File: pkg/envoy/ads/cache_stream.go (hunk 9 above)
Comment (steeling) [reply pkg/envoy/ads/cache_stream.go]: yup, rebased and removed

File: pkg/injector/bootstrap.go (hunk 8 above)
Comment (steeling) [reply pkg/injector/bootstrap.go:41]: I have https://github.com/openservicemesh/osm/pull/5086, happy to submit that first

**Review by shashankram — APPROVED** [review]
